<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Project: Job Analysis from HeadHunter
   

In [3]:
import pandas as pd
import psycopg2

In [ ]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

In [4]:
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    host='84.201.134.129',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    port= 5432
)

# Unit 3: Preliminary Data Analysis

1. Write a query to count the number of job postings in our database (the job postings are stored in the vacancies table). 

In [23]:
# Query text
query_3_1 = f'''select
            count(id)
            from vacancies
            
'''

In [24]:
# Query result
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Write a query to count the number of employers (table employers).

In [28]:
# Query text
query_3_2 = f'''select
            count(id)
            from employers
            
'''

In [31]:
# Query result
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Write a query to count the number of regions (table areas).

In [32]:
# Query text
query_3_3 = f'''select
            count(id)
            from areas
            
'''

In [34]:
# Query result
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4.Write a query to count the number of industries in the database (table industries).

In [35]:
# Query text
query_3_4 = f'''select
            count(id)
            from industries
            
'''

In [36]:
# Query result
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

### Conclusions from the preliminary data analysis

The preliminary data analysis revealed that in 1,362 regions, 23,501 employers from 294 industries published 49,197 job postings.

Of these:

* Top 5 regions: Moscow (5,333), St. Petersburg (2,851), Minsk (2,112), Novosibirsk (2,006), Almaty (1,892).
* Top 5 companies: Yandex (1,933), Rostelecom (491), Tinkoff (444), SBER (428), Gazprom Neft (331).
* Top 5 professions: System Administrator (1,761), 1C Programmer (1,153), Business Analyst (488), Technical Support Specialist (457), System Analyst (432).
 


# Unit 4: Detailed Job Analysis

1. Write a query to find out how many job postings (cnt) there are in each region (area).
Sort the results by the number of job postings in descending order.

In [43]:
# тQuery text
query_4_1 = f'''select
            count(v.id),
            a.name
            from vacancies v
            join areas a on v.area_id = a.id
            group by a.id
            order by 1 desc
            
'''

In [44]:
# Query result
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


count                      name
0     5333                    Москва
1     2851           Санкт-Петербург
2     2112                     Минск
3     2006               Новосибирск
4     1892                    Алматы
..     ...                       ...
764      1                    Кизляр
765      1                    Джизак
766      1                    Эртиль
767      1                  Арсеньев
768      1  Тура (Красноярский край)

[769 rows x 2 columns]

2. Write a query to determine how many job postings have at least one of the two salary fields filled.

In [76]:
# Query text
query_4_2 = f'''select
            count(id)
            from vacancies
            where ((salary_from is not NULL) OR (salary_to is not NULL))
            
            
'''

In [77]:
# Query result
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Find the average values for the lower and upper salary range limits. Round the values to the nearest whole number.

In [84]:
# Query text
query_4_3 = f'''select
            AVG(salary_from),
            AVG(salary_to)
            from vacancies
            
            
            
'''

In [85]:
# Query result
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg            avg
0  71064.657901  110536.741923

4. Write a query to display the number of job postings for each combination of work schedule type (schedule) and employment type (employment) used in the job postings.
Sort the results by the number of postings in descending order.

In [88]:
# Query text
query_4_4 = f'''select
            schedule,
            employment,
            count(id)
            from vacancies
            Group by schedule, employment
            order by 3  desc
            
            
'''

In [89]:
# Query result
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Write a query to display the values of the "Required work experience" (experience) field, ordered by the number of job postings that specify each experience level, in ascending order.

In [98]:
# Query text
query_4_5 = f'''select
            experience
            from vacancies
            group by experience
            order by count(id) asc
            
            
'''

In [99]:
# Query result
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience
0         Более 6 лет
1           Нет опыта
2       От 3 до 6 лет
3  От 1 года до 3 лет

***

### Conclusions from the detailed job analysis

* Most of the job postings are from Moscow-based employers (5,333).

* Employers are primarily looking for full-time specialists (35,367), while those seeking full-time remote workers are five times fewer (7,082). This is likely due to the unpopularity of remote work and the nature of the jobs. For example, a system administrator (the most popular job) cannot always perform their duties remotely. Another factor could be that it is easier to control employees in the office.

* A high number of job postings require 1 to 3 years of experience (26,152), which suggests that employers are willing to consider specialists with minimal experience. These specialists do not demand high salaries and are eager to grow in their profession. On the other hand, specialists with 6+ years of experience (1,337) are less in demand—likely because they already have extensive knowledge, expect higher salaries, and may find it more difficult to adapt to new employers.

* The average salary range: the lower bound is around 71k, and the upper bound is around 110k. This further confirms that employers are looking for specialists with 1 to 3 years of experience who do not have high salary expectations.

# Unit 5: Employer Analysis

1. Write a query to find out which employers are in the first and fifth positions by the number of job postings.

In [102]:
# Query text
query_5_1 = f'''select
            e.name
            from employers e
            join vacancies v on v.employer_id = e.id
            group by e.name
            order by count(v.id) desc
            
            
'''

In [103]:
# Query result
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_13132\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name
0                      Яндекс
1                  Ростелеком
2                    Тинькофф
3                        СБЕР
4               Газпром нефть
...                       ...
14761            им. Тьюринга
14762        LLC Novalab Tech
14763   Компания Мир упаковки
14764  Лифинцев Олег Иванович
14765      ТрансЭнергоХолдинг

[14766 rows x 1 columns]

2. Write a query to display the number of employers and job postings for each region.
Among the regions with no job postings, find the one with the highest number of employers.


In [9]:
# Query text
query_5_2 = f'''select
            a.name,
            count(distinct e.id) as employers_cnt,
            count(distinct v.id) as vacancies_cnt
            from areas a
            left join employers e on a.id = e.area 
            left join vacancies v on a.id = v.area_id
            where v.id is null
            group by a.name
            order by 2 desc
            
            
'''

In [10]:
# Query result
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_31392\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  employers_cnt  vacancies_cnt
0                            Россия            410              0
1                         Казахстан            207              0
2                Московская область             75              0
3                Краснодарский край             19              0
4                          Беларусь             18              0
..                              ...            ...            ...
588                        Кувандык              0              0
589                        Куйбышев              0              0
590  Куйбышево (Ростовская область)              0              0
591                        Култаево              0              0
592                        Кульсары              0              0

[593 rows x 3 columns]

3. For each employer, count the number of regions where they post their job vacancies.
Sort the results in descending order by the count.


In [12]:
# Query text
query_5_3 = f'''select
            e.name,
            count(distinct v.area_id)
            from vacancies v
            join employers e on v.employer_id = e.id
            group by e.name
            order by count(v.area_id) desc
            
            
'''

In [13]:
# Query result
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_26932\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  count
0                                     Яндекс    181
1                                 Ростелеком    152
2                                   Тинькофф     43
3                                       СБЕР     24
4                              Газпром нефть     24
...                                      ...    ...
14761             ПСБ Инновации и Инвестиции      1
14762                             ПСЗ Янтарь      1
14763                                    ПСК      1
14764  Псковский завод механических приводов      1
14765                     ПСК ТатЭнергоСтрой      1

[14766 rows x 2 columns]

4. Write a query to count the number of employers who do not have an industry specified.

In [25]:
# Query text
query_5_4 = f'''select
            count(e.id)
            from employers e
            left join employers_industries i on e.id = i.employer_id
            where i.industry_id is null
            
            
            
'''

In [26]:
# Query result
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_26932\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Write a query to find the name of the company that is in third place in the alphabetical list (by name) of companies that have four industries specified.

In [36]:
# Query text
query_5_5 = f'''select
            e.name
            from employers e
            left join employers_industries i on e.id = i.employer_id
            group by e.name
            having count(i.industry_id) = 4
            order by e.name asc
            offset 2
            limit 1
            
            
            
'''

In [37]:
# Query result
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_26932\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. Use a query to find out how many employers have "Software Development" listed as their industry.

In [45]:
# Query text
query_5_6 = f'''select
            count(e.id)
            from employers e
            left join employers_industries i on e.id = i.employer_id
            left join industries n on n.id = i.industry_id
            where n.name = 'Разработка программного обеспечения'
            
            
            
            
'''

In [46]:
# Query result
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_26932\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. For the company "Yandex," display a list of million-plus cities where the company has job postings, along with the number of job postings in each region. Also, include a "Total" row with the total number of job postings for the company. Sort the results in ascending order by the number of postings.

The list of million-plus cities should be taken [from here](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 



In [81]:
# Code to get the list of million-plus cities
import requests

from bs4 import BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')
table = page.find("table", class_="standard")

# Retrieving the cities list  from the table
cities = []
for row in table.find_all('tr')[1:]:  # Skipping the table header
    city_cell = row.find_all('td')[1]  # Retrieving the cell with the city name
    city = city_cell.text.strip()  # Extracting text from the cell
    cities.append(city)
cities = tuple(cities)
print(cities)



('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [100]:
# Query text
query_5_7 = f'''select
            (a.name,
            count(v.id) as cnt
            from vacancies v
            left join areas a on a.id = v.area_id
            left join employers e on e.id = v.employer_id
            where (a.name IN {cities}) and (e.name = 'Яндекс')
            group by a.name)
            
            Union all
            
            (select
            'total',
            count(v.id)
            from vacancies v
            left join areas a on a.id = v.area_id
            left join employers e on e.id = v.employer_id
            where (a.name IN {cities}) and (e.name = 'Яндекс')
            group by 1)
            
            order by 2 asc
            
            
'''

In [101]:
# Query result
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_26932\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            total  485

***

### Conclusions from the employer analysis

From the employer analysis, we found the following:

* The top five employers posting job vacancies.
* In first place is Yandex with 485 job postings across 181 regions.
* There are 410 employers in Russia who currently have no job postings but are registered in the system.

# Unit 6: Subject Matter Analysis

1. How many job postings are related to data?

We consider a job posting to be related to data if its title contains the words "data" or "данн."

Hint: Keep in mind that job titles may be written in any case.


In [108]:
# Query text
query_6_1 = f'''select
            count(id) 
            from vacancies
            where (lower(name) like '%data%') or (lower(name) like '%данн%')
            
            
'''

In [109]:
# Query result
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_26932\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. How many job postings are suitable for a Junior Data Scientist?
We will consider job postings for Data Scientists if their title contains at least one of the following combinations:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (excluding HTML-related positions)
* 'machine learning'
* 'машинн%обучен%'

** In the following tasks, we will continue working with job postings that meet this condition..*

We will consider job postings for Junior specialists if:
* the title contains the word 'junior' *or*
* the required experience is 'No experience' *or*
* the employment type is 'Internship'.
 

In [168]:
# Query text
query_6_2 = f'''select
            count(id) 
            from vacancies
            where ((name ILIKE '%data scientist%') or (name ILIKE '%data science%') 
            or (name ILIKE '%исследователь данных%') or (name like '%ML%') and (name not like '%HTML%') or (name ILIKE '%machine learning%') or (name ILIKE '%машинн%обучен%'))
            and ((name ILIKE '%junior%') or (experience = 'Нет опыта') or (employment = 'Стажировка'))
            
            
'''

In [169]:
# Query result
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_26932\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. How many job postings are there for DS where SQL or Postgres is listed as a key skill?

** The criteria for classifying a job posting as DS are provided in the previous task.*

In [202]:
# Query text
query_6_3 = f'''select 
                count(v.id)
                from vacancies v
                WHERE ( (v.key_skills like '%SQL%')
                or
                (v.key_skills ilike '%postgres%')
                )
                AND
                (
                (name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
                OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
                OR (name ILIKE '%machine learning%')
                OR (name ILIKE '%машинн%обучен%')
                AND (key_skills IS NOT NULL)
                )
            
            
'''            
            

In [203]:
# Query result
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_26932\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Check how popular Python is in employers' requirements for DS. To do this, calculate the number of job postings where Python is listed as a key skill.

** This can be done using a query similar to the previous one.*

In [204]:
# Query text
query_6_4 = f'''select
            count(id) 
            from vacancies
            where ((name ILIKE '%data scientist%') or (name ILIKE '%data science%') 
            or (name ILIKE '%исследователь данных%') or (name ILIKE '%ML%') and (name not ILIKE '%HTML%') or (name ILIKE '%machine learning%') or (name ILIKE '%машинн%обучен%'))
            and (key_skills ILIKE '%Python%')
            
''' 

In [205]:
# Query result
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_26932\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    357

5. On average, how many key skills are listed in job postings for DS?
Round the answer to two decimal places.

In [196]:
# Query text
query_6_5 = f'''select
            round(avg(length(key_skills)- length(replace(key_skills, CHR(9), ''))+1), 2)
            from vacancies
            where ((name ILIKE '%data scientist%') or (name ILIKE '%data science%') 
            or (name ILIKE '%исследователь данных%') or (name like '%ML%') and (name not like '%HTML%') or (name ILIKE '%machine learning%') or (name ILIKE '%машинн%обучен%'))
            and (key_skills is not null)
            
''' 

In [197]:
# Query result
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_26932\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. мWrite a query to calculate **the average salary** indicated for DS positions for each required experience level (unique values from the experience field). 

When solving this task, consider the following:
1. Only include job postings where at least one of the two salary fields is filled.
2. If both salary fields are filled, calculate the salary for each job posting as the sum of the two fields divided by 2. If only one field is filled, use that value as the salary for the job posting.
3. If any calculations involve null, the result will also be null (see what the query select 1 + null returns). To avoid this, use the function [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull),  which replaces null with a value you specify. For example, see what the query 'select 1 + coalesce(null, 0)' returns. 

Determine the average salary that a data scientist with 3 to 6 years of experience can expect. Round the result to the nearest whole number. 

In [212]:
# Query text
query_6_6 = f'''select
            experience, 
            ROUND(AVG(coalesce((salary_from + salary_to) / 2, salary_from, salary_to)), 0)
            from vacancies
            where ((name ILIKE '%data scientist%') or (name ILIKE '%data science%') 
            or (name ILIKE '%исследователь данных%') or (name like '%ML%') and (name not like '%HTML%') or (name ILIKE '%machine learning%') or (name ILIKE '%машинн%обучен%'))
            and (key_skills is not null)
            GROUP BY experience
            ORDER BY 1
            
''' 

In [213]:
# Query result
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Евгения\AppData\Local\Temp\ipykernel_26932\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience     round
0         Более 6 лет       NaN
1           Нет опыта   74643.0
2  От 1 года до 3 лет  146704.0
3       От 3 до 6 лет  243115.0

***

### Conclusions from Subject Matter Analysis

The subject-specific analysis of job postings revealed the following:

* 1,771 job postings are related to data in some way.
* 51 job postings are suitable for entry-level data scientists with no prior work experience.
* 229 job postings list SQL or PostgreSQL as key skills.
* 357 job postings require proficiency in Python.
* Employers list an average of more than 6 skills in job postings for data scientists.
* The salary for data science professionals ranges from 74,000 RUB for entry-level specialists to 243,000 RUB for specialists with 3 to 6 years of experience.

# Overall Project Conclusion

The conducted job analysis allowed us to identify the main requirements employers have for candidates.
We determined the largest employer—Yandex, as well as the most active region—Russia. We examined the key skills required by employers, which include not only hard skills but also soft skills (teamwork, responsibility, etc.). Additionally, we identified the industry with the highest number of job openings.